middleware/instrument_test.go

```diff
@@ -0,0 +1,148 @@
+			`
+			}
+
+			require.NoError(t, testutil.GatherAndCompare(reg, strings.NewReader(output), "slow_request_throughput_"+i.ThroughputUnit))
+		})
+	}
+}
+
+func NewInstrument(registry *prometheus.Registry) Instrument {
+	reg := promauto.With(registry)
+
+	const metricsNativeHistogramFactor = 1.1
+	const throughputUnit = "unit"
+	const SlowRequestCutoff = 100 * time.Millisecond
+
+	return Instrument{
+		Duration: reg.NewHistogramVec(prometheus.HistogramOpts{
+			Name:                            "request_duration_seconds",
+			Help:                            "Time (in seconds) spent serving HTTP requests.",
+			Buckets:                         instrument.DefBuckets,
+			NativeHistogramBucketFactor:     metricsNativeHistogramFactor,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: time.Hour,
+		}, []string{"method", "route", "status_code", "ws"}),
+		PerTenantDuration: reg.NewHistogramVec(prometheus.HistogramOpts{
+			Name:                            "per_tenant_request_duration_seconds",
+			Help:                            "Time (in seconds) spent serving HTTP requests for a particular tenant.",
+			Buckets:                         instrument.DefBuckets,
+			NativeHistogramBucketFactor:     metricsNativeHistogramFactor,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: time.Hour,
+		}, []string{"method", "route", "status_code", "ws", "tenant"}),
+		RequestBodySize: reg.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "request_message_bytes",
+			Help:    "Size (in bytes) of messages received in the request.",
+			Buckets: BodySizeBuckets,
+		}, []string{"method", "route"}),
+		ResponseBodySize: reg.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "response_message_bytes",
+			Help:    "Size (in bytes) of messages sent in response.",
+			Buckets: BodySizeBuckets,
+		}, []string{"method", "route"}),
+		InflightRequests: reg.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "inflight_requests",
+			Help: "Current number of inflight requests.",
+		}, []string{"method", "route"}),
+		RequestCutoff:  SlowRequestCutoff,
+		ThroughputUnit: throughputUnit,
+		RequestThroughput: reg.NewHistogramVec(prometheus.HistogramOpts{
+			Name:                            "request_throughput_" + throughputUnit,
+			Help:                            "Server throughput running requests.",
+			ConstLabels:                     prometheus.Labels{"cutoff_ms": strconv.FormatInt(SlowRequestCutoff.Milliseconds(), 10)},
+			Buckets:                         []float64{1, 5, 10},
+			NativeHistogramBucketFactor:     metricsNativeHistogramFactor,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: time.Hour,
+		}, []string{"method", "route"}),
+	}
+}
+
+func TestExtractValueFromMultiValueHeader(t *testing.T) {
+	tests := []struct {
+		header   string
+		key      string
+		expected int64
+		err      bool
+	}{
+		{"key0=0, key1=1", "key0", 0, false},
```

Comment:
can you add keys to each struct initialization? 
```suggestion
		{header: "key0=0, key1=1", key: "key0", expected: 0, err: false},
```

maybe putting them on separate lines might make this more readable


```suggestion
		{
		   header: "key0=0, key1=1",
		   key: "key0",
		   expected: 0,
		   err: false,
		 },
```